base/shared/src/main/scala/scalaz/data/IStream.scala

```diff
@@ -0,0 +1,19 @@
+package scalaz
+package data
+
+trait IStreamModule {
+  type IStream[A]
+
+  def scons[A](a: A)(as: IStream[A]): IStream[A]
+  def uncons[A](as: IStream[A]): (A, IStream[A])
+}
+
+private[data] object IStreamImpl extends IStreamModule {
+  type IStream[A] = data.Cofree[Id, A]
```

Comment:
With the current definitions of `Cofree` and `EnvT`, this looks like constructing an infinite stream eagerly.